core/src/main/java/com/linecorp/armeria/server/healthcheck/CpuHealthChecker.java

```diff
@@ -0,0 +1,180 @@
+    }
+
+    private CpuHealthChecker(double targetSystemCpuUsage, double targetProcessCpuLoad,
+                             DoubleSupplier systemCpuUsageSupplier, DoubleSupplier processCpuUsageSupplier) {
+        checkArgument(targetSystemCpuUsage >= 0 && targetSystemCpuUsage <= 1.0,
+                      "cpuUsage: %s (expected: 0 <= cpuUsage <= 1)", targetSystemCpuUsage);
+        checkArgument(targetProcessCpuLoad >= 0 && targetProcessCpuLoad <= 1.0,
+                      "processCpuLoad: %s (expected: 0 <= processCpuLoad <= 1)", targetProcessCpuLoad);
+        this.targetSystemCpuUsage = targetSystemCpuUsage;
+        this.targetProcessCpuLoad = targetProcessCpuLoad;
+        this.systemCpuUsageSupplier = systemCpuUsageSupplier;
+        this.processCpuUsageSupplier = processCpuUsageSupplier;
+        checkState(operatingSystemBeanClass != null, "Unable to find an 'OperatingSystemMXBean' class");
+        checkState(operatingSystemBean != null, "Unable to find an 'OperatingSystemMXBean'");
+        checkState(systemCpuLoad != null, "Unable to find the method 'OperatingSystemMXBean.getCpuLoad'" +
+                                           " or 'OperatingSystemMXBean.getSystemCpuLoad'");
+        checkState(processCpuLoad != null,
+                   "Unable to find the method 'OperatingSystemMXBean.getProcessCpuLoad'");
+    }
+
+    private static double invoke(@Nullable MethodHandle mh) {
+        if (mh == null) {
+            return Double.NaN;
+        }
+
+        try {
+            return (double) mh.invoke(operatingSystemBean);
+        } catch (Throwable e) {
+            return Double.NaN;
+        }
+    }
+
+    @Nullable
+    @SuppressWarnings("ReturnValueIgnored")
+    private static MethodHandle detectMethod(final String name) {
+        if (operatingSystemBeanClass == null) {
+            return null;
+        }
+        try {
+            // ensure the Bean we have is actually an instance of the interface
+            operatingSystemBeanClass.cast(operatingSystemBean);
+            final Method method = operatingSystemBeanClass.getMethod(name);
+            return LOOKUP.unreflect(method);
+        } catch (ClassCastException | NoSuchMethodException | SecurityException | IllegalAccessException e) {
+            logger.warn("Failed to detect method {}.{} for {}", operatingSystemBeanClass.getSimpleName(),
+                        name, CpuHealthChecker.class.getSimpleName(), e);
+            return null;
+        }
+    }
+
+    @Nullable
+    private static Class<?> getFirstClassFound(final List<String> classNames) {
+        for (String className : classNames) {
+            try {
+                return Class.forName(className, false, CpuHealthChecker.class.getClassLoader());
+            } catch (ClassNotFoundException ignore) {
+            }
+        }
+        return null;
+    }
+
+    static CpuHealthChecker of(double targetCpuUsage, double targetProcessCpuLoad) {
```

Comment:
What do you think of:
1. Moving this to `HealthChecker.java`
2. Renaming this method from `of` to `ofCpu`
3. Making this method public so that users can actually use this class
4. Adding javadocs explaining what this health checker does, and what each parameter does

Reply:
I saw this comment late:

https://github.com/line/armeria/pull/4673#discussion_r1502153234

I'm fine if this is handled separately, I would just like to note that users won't be able to access this method since it is not `public`